test/test_torch.py

```diff
@@ -17224,6 +17224,76 @@ def test_reciprocal_complex_extremal(self, device, dtype):
 
         self.compare_with_numpy(torch.reciprocal, np.reciprocal, vals, device, dtype)
 
+    def test_copysign_bool(self, device):
+        # Promoted to float
+        # Only 0 and 1
+        x = torch.tensor([0, 0, 1, 1], device=device, dtype=torch.bool)
+        y = torch.tensor([0, 1, 0, 1], device=device, dtype=torch.bool)
+        res = torch.tensor([0, 0, 1, 1], device=device, dtype=torch.float)
+        self.assertEqual(torch.copysign(x, y), res)
+
+    def _test_copysign(self, device, dtype, result_dtype):
+        # Scalar
+        x = torch.tensor([-1, 0, -0, 1], dtype=dtype)
+        y = torch.tensor(-1.)
+        res = torch.tensor([-1, -0, -0, -1], dtype=result_dtype)
+        self.assertEqual(torch.copysign(x, y), res)
+
+        x = torch.tensor(-1.)
+        y = torch.tensor([-1, 0, -0, 1], dtype=dtype)
+        res = torch.tensor([-1, 1, 1, 1], dtype=result_dtype)
+        self.assertEqual(torch.copysign(x, y), res)
+
+        # Constant
+        x = torch.tensor([-1, 0, -0, 1], dtype=dtype)
+        res = torch.tensor([-1, -0, -0, -1], dtype=result_dtype)
+        self.assertEqual(torch.copysign(x, -1.), res)
+
+        # Normal
+        res = torch.tensor([[1, 0, 0, 1] * 2, [-1, -0, -0, -1] * 2], dtype=result_dtype)
+        x = torch.tensor([-1, -0, 0, 1] * 2, dtype=dtype)
+        y = torch.tensor([[0] * 4, [1] * 4], dtype=dtype).reshape(-1)
+        self.assertEqual(torch.copysign(x, y), res[0])
+
+        y = torch.copysign(y, -1)
+        self.assertEqual(torch.copysign(x, y), res[1])
+
+        # Broadcast
+        res = torch.tensor([[[1, 0, 0, 1]] * 2, [[-1, -0, -0, -1]] * 2], dtype=result_dtype)
+
+        # LHS
+        x = torch.tensor([-1, -0, 0, 1] * 2, dtype=dtype).reshape(2, 1, 4)
+        y = torch.tensor([[0] * 4, [1] * 4], dtype=dtype)
+        z = torch.copysign(x, y)
+        for i in range(2):
+            self.assertEqual(z[i], res[0])
+
+        y = torch.copysign(y, -1)
+        z = torch.copysign(x, y)
+        for i in range(2):
+            self.assertEqual(z[i], res[1])
+
+        # RHS
+        x = torch.tensor([-1, 0, -0, 1] * 2, dtype=dtype).reshape(2, 4)
+        y = torch.tensor([[0] * 4, [1] * 4], dtype=dtype).reshape(2, 1, 4)
+        z = torch.copysign(x, y)
+        for i in range(2):
+            self.assertEqual(z[i], res[0])
+
+        y = torch.copysign(y, -1)
+        z = torch.copysign(x, y)
+        for i in range(2):
+            self.assertEqual(z[i], res[1])
```

Comment:
nit: One thing we can consider doing is comparing the result of torch.copysign to numpy.copysign instead of coming up with these test cases. For example, we could write a 
```
def _compare_with_numpy(a, b):
    torch_result = torch.copysign(a, b)

    # To handle inconsistencies in our type promotion and numpy's type promotion
    expected_dtype = torch.promote_types(a, b)
    expected = torch.from_numpy(np.copysign(a.cpu().numpy(), b.cpu().numpy()))
    expected.to(expected_dtype)
```
and then call `_compare_with_numpy(a, b)` on some different tensors `a` and `b`. I'm not actually sure if the sample code handles all of the type promotion edge cases, because there is going to be different behavior between -0.0 and -0 due to how -0 doesn't actually exist.